otx/algorithms/detection/tasks/train.py

```diff
@@ -0,0 +1,265 @@
+
+# Copyright (C) 2022 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+
+import io
+from typing import Any, DefaultDict, Iterable, List, Optional
+
+import numpy as np
+import torch
+from mmcv.utils import ConfigDict
+from mpa.utils.logger import get_logger
+
+from otx.algorithms.common.adapters.mmcv.hooks import OTXLoggerHook
+from otx.algorithms.detection.utils.otx_utils import TrainingProgressCallback
+from otx.api.configuration import cfg_helper
+from otx.api.configuration.helper.utils import ids_to_strings
+from otx.api.entities.datasets import DatasetEntity
+from otx.api.entities.inference_parameters import InferenceParameters
+from otx.api.entities.metrics import (
+    BarChartInfo,
+    BarMetricsGroup,
+    CurveMetric,
+    LineChartInfo,
+    LineMetricsGroup,
+    MetricsGroup,
+    ScoreMetric,
+    VisualizationType,
+)
+from otx.api.entities.model import ModelEntity
+from otx.api.entities.resultset import ResultSetEntity
+from otx.api.entities.subset import Subset
+from otx.api.entities.train_parameters import TrainParameters, default_progress_callback
+from otx.api.serialization.label_mapper import label_schema_to_bytes
+from otx.api.usecases.evaluation.metrics_helper import MetricsHelper
+from otx.api.usecases.tasks.interfaces.training_interface import ITrainingTask
+
+from .inference import DetectionInferenceTask
+
+logger = get_logger()
+
+
+# pylint: disable=too-many-locals, too-many-instance-attributes
+class DetectionTrainTask(DetectionInferenceTask, ITrainingTask):
+    """Train Task Implementation of OTX Detection."""
+
+    def save_model(self, output_model: ModelEntity):
+        """Save best model weights in DetectionTrainTask."""
+        logger.info("called save_model")
+        buffer = io.BytesIO()
+        hyperparams_str = ids_to_strings(cfg_helper.convert(self._hyperparams, dict, enum_to_str=True))
+        labels = {label.name: label.color.rgb_tuple for label in self._labels}
+        model_ckpt = torch.load(self._model_ckpt)
+        modelinfo = {
+            "model": model_ckpt["state_dict"],
+            "config": hyperparams_str,
+            "labels": labels,
+            "confidence_threshold": self.confidence_threshold,
+            "VERSION": 1,
+        }
+        if (
+            self._model_cfg
+            and hasattr(self._model_cfg.model, "bbox_head")
+            and hasattr(self._model_cfg.model.bbox_head, "anchor_generator")
+        ):
+            if getattr(
```

Comment:
how about combining this line with the line right above it?

Reply:
Edited.

Reply:
I'm sorry I forgot the word combine. isn't it better to combine all hasattr lines together?